californium-core/src/main/java/org/eclipse/californium/core/network/TcpMatcher.java

```diff
@@ -47,29 +50,42 @@
 	 * Creates a new matcher for running CoAP over TCP.
 	 * 
 	 * @param config the configuration to use.
-	 * @throws NullPointerException if the configuration is {@code null}.
+	 * @param notificationListener the callback to invoke for notifications
+	 *            received from peers.
+	 * @param observationStore the object to use for keeping track of
+	 *            observations created by the endpoint this matcher is part of.
+	 * @throws NullPointerException if the configuration, notification listener,
+	 *             or the observation store is {@code null}.
 	 */
-	public TcpMatcher(final NetworkConfig config) {
-		super(config);
+	public TcpMatcher(final NetworkConfig config, final NotificationListener notificationListener,
+			final ObservationStore observationStore) {
+		super(config, notificationListener, observationStore);
 	}
 
-	@Override public void sendRequest(Exchange exchange, Request request) {
+	@Override
+	public void sendRequest(Exchange exchange, final Request request) {
 
 		if (request.getToken() != null) {
 			KeyToken idByToken = Exchange.KeyToken.fromOutboundMessage(request);
 			// ongoing requests may reuse token
-			if (!(exchange.getFailedTransmissionCount() > 0 || request.getOptions().hasBlock1() || request.getOptions()
-					.hasBlock2() || request.getOptions().hasObserve()) && exchangeStore.get(idByToken) != null) {
+			if (!(exchange.getFailedTransmissionCount() > 0 || request.getOptions().hasBlock1()
```

Comment:
The whole `if (request.getToken() != null)` block should be removed because this check is now done in `InMemoryMessageExchangeStore`.